translate/pootle
# Cache virtual folder stats using CachedTreeItem

@translate/core  Ready to review.

The purpose of these changes is to use CachedTreeItem in order to keep stats in cache, resulting in a faster retrieval. In order to do so VFolderCachedTreeItem trees are created besides regular tree, see an example in https://imgrush.com/leOJqX9pAlUj. Unfortunately creating those tree structures results in worsening the vfolder and regular directory name clash (#3534) which was reported as #3809. Stats can only be calculated using `refresh_stats_rq`.

Instructions for setup:

- Check out the branch locally
- You might need to make the assets to ensure you use the latest JS.
- Run `./manage.py shell_plus`
- Run `VirtualFolder.objects.all().delete()` and exit the shell
- If you have a `test_vfolders` project remove it from Pootle and from `POOTLE_TRANSLATION_DIRECTORY`
- Get https://docs.google.com/file/d/0B-8KVyCPnkZRSHAzbUFRYmV3LWM and uncompress it on your `POOTLE_TRANSLATION_DIRECTORY`
- Add a new `test_vfolders` project.
- Run `./manage.py update_stores --project=test_vfolders`
- Get http://dpaste.com/15NCKRS and save it as `vfolders_test_directory.json`
- Run `./manage.py add_vfolders vfolders_test_directory.json`
- Run `./manage.py refresh_stats_rq --project=test_vfolders`


Instructions for testing:

- Run `./manage.py runserver`
- Log in as a regular user (not admin)
- Go to `http://127.0.0.1:8000/af/test_vfolders/browser/chrome/`
- Check that:
  - the vfolders stats table is displayed
  - the listed vfolders are all the vfolders that should be displayed. The following must not be displayed:
    - not browsable vfolders,
    - completely translated vfolders,
    - priority below 1 vfolders,
    - vfolders that don't match any file in that path.
  - you can see completely translated vfolders that have failing checks.
  - the counts match. Note that each vfolder counts must only include the files they match in the current location or below it.
  -  the translate links are correct. Just are like regular translate links for this location, but with the vfolder name included just after its location, for example a vfolder with location `/af/test_vfolders/browser/` and name `test-overlapping-vfolder` will have a needs work translate link pointing to `http://127.0.0.1:8000/af/test_vfolders/translate/browser/test-overlapping-vfolder/chrome/#filter=incomplete`
- Log in as an admin user
- Check that in `http://127.0.0.1:8000/af/test_vfolders/browser/chrome/` now you can also see styled as disabled (grayed out):
  - not browsable vfolders ,
  - completely translated vfolders,
  - vfolders with priority below 1.
- Click on the `need translation` translate link for vfolder `test-overrides` on the vfolders stats table and ensure that:
  - you see `browser/test-overrides/chrome` in the navbar dropdown,
  - the URL in the browser location bar is `http://127.0.0.1:8000/af/test_vfolders/translate/browser/test-overrides/chrome/#filter=incomplete`,
  - the units displayed only belong to files matched by the vfolder filter in `/af/test_vfolders/browser/chrome/`
  - it is not possible to sort by priority.
- Select the same path to be translated in the editor, but not restricted to a vfolder: `http://127.0.0.1:8000/af/test_vfolders/translate/browser/chrome/#filter=incomplete`. Check that:
  - the unit count is bigger
  - you can sort by priority.
- Open the navbar dropdown and ensure that all the vfolders that match within current project are displayed:
  - Should be included (even if they have low priority or are completely translated):
    - test-overlapping-vfolder
    - test-overrides
    - test-low-priority
    - test-vfolder1
    - test-vfolder2
    - test-vfolder3
    - test-completely-translated-with-failing-checks
  - Are not browsable and thus should not be included:
    - test-not-browsable
    - test-completely-translated
  - Don't apply to any file, so they should not be included:
    - test-directory
    - test-newfile
- Open `http://127.0.0.1:8000/af/test_vfolders/translate/chat/#filter=incomplete` and ensure that:
  - no priority is being displayed for any of the units listed,
  - it is not possible to select priority sorting.
- Open `http://127.0.0.1:8000/af/test_vfolders/translate/dom/chrome/#filter=incomplete` and ensure that:
  - the top priority is being displayed for all the units shown (the units in this location belong to the vfolders whose name starts with `test-vfolder`, which eases looking in the JSON). Remember to ensure that the top priority is shown for a unit that is included in several vfolders.
  - you can sort by priority. Remember top priorities are above and lower at the bottom.
